### PR TITLE
Fixes #29075 - add pulp_rpm plugin

### PR DESCRIPTION
**Pull request by wbclark**

### CODE REVIEW

**Comment by wbclark** [comment]
As expected at this time, it's failing to find a package to install:

```
Error: Execution of '/usr/bin/yum -d 0 -e 0 -y install python3-pulp-pulp_rpm' returned 1: Error: Nothing to do

  Error: /Stage[main]/Pulpcore::Plugin::Rpm/Pulpcore::Plugin[pulp_rpm]/Package[python3-pulp-pulp_rpm]/ensure: change from 'purged' to 'present' failed: Execution of '/usr/bin/yum -d 0 -e 0 -y install python3-pulp-pulp_rpm' returned 1: Error: Nothing to do
```

Although the package name is probably not correct, but I don't find under any other name in http://koji.katello.org/releases/yum/katello-nightly/pulpcore/el7/x86_64/ either.

@evgeni could you direct me where to find it, if it exists at this point? I did see https://github.com/theforeman/foreman-packaging/pull/4191 was merged, hence my confusion on the topic.

Thanks!

**Comment by evgeni** [comment]
The python plugin was not packaged yet

File: manifests/plugin/rpm.pp [hunk 1]
```diff
@@ -0,0 +1,5 @@
+# @summary Pulp RPM plugin
+class pulpcore::plugin::rpm {
+  pulpcore::plugin { 'pulp_rpm':
```
Comment (ehelms) [inline manifests/plugin/rpm.pp:3]: Looking at http://koji.katello.org/releases/yum/katello-nightly/pulpcore/el8/x86_64/python3-pulp-rpm-3.3.1-1.el8.noarch.rpm -- I think this should just be `rpm`

**Comment by wbclark** [comment]
I updated to have the correct plugin/package name but I'm still anticipating a failure based on what I observed when I tried manually installing it.

File: Gemfile [hunk 2]
```diff
@@ -5,6 +5,7 @@ source 'https://rubygems.org'
 
 gem 'puppet', ENV.key?('PUPPET_VERSION') ? "~> #{ENV['PUPPET_VERSION']}" : '>= 5.5'
 
+gem 'beaker', '< 4.22.0'
```
Comment (ekohl) [inline Gemfile:8]: Temporary workaround that can be dropped.

**Comment by ehelms** [comment]
Looks like now it's having issues on EL7 and EL8 with finding solv even though that is packaged and in the repository:

```
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[python3-django-admin migrate --noinput]/returns: pkg_resources.DistributionNotFound: The 'solv' distribution was not found and is required by the application
  Info: Pulpcore::Admin[reset-admin-password --random]: Unscheduling all events on Pulpcore::Admin[reset-admin-password --random]
  Error: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[python3-django-admin migrate --noinput]: Failed to call refresh: 'python3-django-admin migrate --noinput' returned 1 instead of one of [0]
  Error: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[python3-django-admin migrate --noinput]: 'python3-django-admin migrate --noinput' returned 1 instead of one of [0]
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[reset-admin-password --random]/Exec[python3-django-admin reset-admin-password --random]: Dependency Exec[python3-django-admin migrate --noinput] has failures: true
  Warning: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[reset-admin-password --random]/Exec[python3-django-admin reset-admin-password --random]: Skipping because of failed dependencies
```

I'll try to debug it.

**Comment by evgeni** [comment]
When you start a python3 shell, can you run the following:
```python

import pkg_resources
import solv

pkg_resources.get_distribution('solv')
```
And post the output?

**Comment by ehelms** [comment]
Looks similar:

```
[root@centos8 vagrant]# python3
Python 3.6.8 (default, May 21 2019, 23:51:36) 
[GCC 8.2.1 20180905 (Red Hat 8.2.1-3)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pkg_resources
>>> import solv
>>> 
>>> pkg_resources.get_distribution('solv')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 468, in get_distribution
    dist = get_provider(dist)
  File "/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 344, in get_provider
    return working_set.find(moduleOrReq) or require(str(moduleOrReq))[0]
  File "/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 888, in require
    needed = self.resolve(parse_requirements(requirements))
  File "/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 774, in resolve
    raise DistributionNotFound(req, requirers)
pkg_resources.DistributionNotFound: The 'solv' distribution was not found and is required by the application
```

**Comment by ehelms** [comment]
If it helps:

```
[root@centos8 vagrant]# rpm -q python3-solv --list
/usr/lib/.build-id
/usr/lib/.build-id/c8
/usr/lib/.build-id/c8/f91b58a1c9bd8367b3ba20a7f5df00fc90c144
/usr/lib64/python3.6/site-packages/__pycache__/solv.cpython-36.opt-1.pyc
/usr/lib64/python3.6/site-packages/__pycache__/solv.cpython-36.pyc
/usr/lib64/python3.6/site-packages/_solv.so
/usr/lib64/python3.6/site-packages/solv.py
```

**Comment by evgeni** [comment]
> If it helps:
> 
> ```
> [root@centos8 vagrant]# rpm -q python3-solv --list
> /usr/lib/.build-id
> /usr/lib/.build-id/c8
> /usr/lib/.build-id/c8/f91b58a1c9bd8367b3ba20a7f5df00fc90c144
> /usr/lib64/python3.6/site-packages/__pycache__/solv.cpython-36.opt-1.pyc
> /usr/lib64/python3.6/site-packages/__pycache__/solv.cpython-36.pyc
> /usr/lib64/python3.6/site-packages/_solv.so
> /usr/lib64/python3.6/site-packages/solv.py
> ```

Indeed. It's missing the egg files, which seem not to be built by cmake for some reason.

**Comment by ehelms** [comment]
@evgeni another related issue to how the python package was created?

```
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[python3-django-admin migrate --noinput]/returns: Traceback (most recent call last):
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[python3-django-admin migrate --noinput]/returns:   File "/usr/bin/python3-django-admin", line 11, in <module>
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[python3-django-admin migrate --noinput]/returns:     load_entry_point('Django==2.2.12', 'console_scripts', 'django-admin')()
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[python3-django-admin migrate --noinput]/returns:   File "/usr/lib/python3.6/site-packages/django/core/management/__init__.py", line 381, in execute_from_command_line
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[python3-django-admin migrate --noinput]/returns:     utility.execute()
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[python3-django-admin migrate --noinput]/returns:   File "/usr/lib/python3.6/site-packages/django/core/management/__init__.py", line 357, in execute
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[python3-django-admin migrate --noinput]/returns:     django.setup()
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[python3-django-admin migrate --noinput]/returns:   File "/usr/lib/python3.6/site-packages/django/__init__.py", line 24, in setup
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[python3-django-admin migrate --noinput]/returns:     apps.populate(settings.INSTALLED_APPS)
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[python3-django-admin migrate --noinput]/returns:   File "/usr/lib/python3.6/site-packages/django/apps/registry.py", line 122, in populate
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[python3-django-admin migrate --noinput]/returns:     app_config.ready()
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[python3-django-admin migrate --noinput]/returns:   File "/usr/lib/python3.6/site-packages/pulpcore/app/apps.py", line 81, in ready
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[python3-django-admin migrate --noinput]/returns:     self.import_viewsets()
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[python3-django-admin migrate --noinput]/returns:   File "/usr/lib/python3.6/site-packages/pulpcore/app/apps.py", line 115, in import_viewsets
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[python3-django-admin migrate --noinput]/returns:     self.viewsets_module = import_module(viewsets_module_name)
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[python3-django-admin migrate --noinput]/returns:   File "/usr/lib64/python3.6/importlib/__init__.py", line 126, in import_module
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[python3-django-admin migrate --noinput]/returns:     return _bootstrap._gcd_import(name[level:], package, level)
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[python3-django-admin migrate --noinput]/returns:   File "<frozen importlib._bootstrap>", line 994, in _gcd_import
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[python3-django-admin migrate --noinput]/returns:   File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[python3-django-admin migrate --noinput]/returns:   File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[python3-django-admin migrate --noinput]/returns:   File "<frozen importlib._bootstrap>", line 665, in _load_unlocked
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[python3-django-admin migrate --noinput]/returns:   File "<frozen importlib._bootstrap_external>", line 678, in exec_module
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[python3-django-admin migrate --noinput]/returns:   File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[python3-django-admin migrate --noinput]/returns:   File "/usr/lib/python3.6/site-packages/pulp_2to3_migration/app/viewsets.py", line 22, in <module>
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[python3-django-admin migrate --noinput]/returns:     from .serializers import (
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[python3-django-admin migrate --noinput]/returns:   File "/usr/lib/python3.6/site-packages/pulp_2to3_migration/app/serializers.py", line 10, in <module>
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[python3-django-admin migrate --noinput]/returns:     from pulp_2to3_migration.app.plugin import PLUGIN_MIGRATORS
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[python3-django-admin migrate --noinput]/returns:   File "/usr/lib/python3.6/site-packages/pulp_2to3_migration/app/plugin/__init__.py", line 16, in <module>
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[python3-django-admin migrate --noinput]/returns:     PLUGIN_MIGRATORS[entry_point.name] = entry_point.load()
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[python3-django-admin migrate --noinput]/returns:   File "/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2318, in load
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[python3-django-admin migrate --noinput]/returns:     return self.resolve()
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[python3-django-admin migrate --noinput]/returns:   File "/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2324, in resolve
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[python3-django-admin migrate --noinput]/returns:     module = __import__(self.module_name, fromlist=['__name__'], level=0)
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[python3-django-admin migrate --noinput]/returns:   File "/usr/lib/python3.6/site-packages/pulp_2to3_migration/app/plugin/rpm/migrator.py", line 21, in <module>
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[python3-django-admin migrate --noinput]/returns:     from pulp_rpm.app.tasks.synchronizing import RpmContentSaver
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[python3-django-admin migrate --noinput]/returns:   File "/usr/lib/python3.6/site-packages/pulp_rpm/app/tasks/__init__.py", line 2, in <module>
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[python3-django-admin migrate --noinput]/returns:     from .synchronizing import synchronize  # noqa
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[python3-django-admin migrate --noinput]/returns:   File "/usr/lib/python3.6/site-packages/pulp_rpm/app/tasks/synchronizing.py", line 68, in <module>
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[python3-django-admin migrate --noinput]/returns:     from pulp_rpm.app.modulemd import (
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[python3-django-admin migrate --noinput]/returns:   File "/usr/lib/python3.6/site-packages/pulp_rpm/app/modulemd.py", line 9, in <module>
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[python3-django-admin migrate --noinput]/returns:     gi.require_version('Modulemd', '2.0')
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[python3-django-admin migrate --noinput]/returns:   File "/usr/lib64/python3.6/site-packages/gi/__init__.py", line 99, in require_version
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[python3-django-admin migrate --noinput]/returns:     raise ValueError('Namespace %s not available' % namespace)
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[python3-django-admin migrate --noinput]/returns: ValueError: Namespace Modulemd not available
  Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[reset-admin-password --random]/Exec[python3-django-admin reset-admin-password --random]: Dependency Exec[python3-django-admin migrate --noinput] has failures: true
```

**Comment by evgeni** [comment]
@ehelms we're probably missing a `Requires: libmodulemd` or so, will look into it.

**Comment by evgeni** [comment]
Oh, no, our libmodulemd is too old. Fuck.

**Comment by jlsherrill** [comment]
The libmodulemd issue is now resolved, right @evgeni ?  https://github.com/theforeman/foreman-packaging/commit/764bec8b47789eb215d5f1d9ca2a77065ab3928d

**Comment by evgeni** [comment]
> The libmodulemd issue is now resolved, right @evgeni ? [theforeman/foreman-packaging@764bec8](https://github.com/theforeman/foreman-packaging/commit/764bec8b47789eb215d5f1d9ca2a77065ab3928d)

no, that's the "solv" fix, not the "modulemd" fix. you can watch me throwing tables on twitch if you like :)

**Comment by evgeni** [comment]
modulemd might be fixed by https://github.com/theforeman/foreman-packaging/pull/5355

@ehelms can I easily test that with a scratch build?

**Comment by evgeni** [comment]
alternative to test this, w/o the fixed build: install libmodulemd (on el8) or libmodulemd2 (on el7) manually/via puppet

**Comment by evgeni** [comment]
btw, I don't understand the EL8 failure:

```
pkg_resources.DistributionNotFound: The 'libcomps~=0.1.11' distribution was not found and is required by the application
```

but

```python
>>> import pkg_resources
>>> pkg_resources.get_distribution('libcomps~=0.1.11')
libcomps 0.1.15 (/usr/lib64/python3.6/site-packages)
```

**Comment by evgeni** [comment]
It tells us that python3-libcomps is not installed, which is not possible?

**Comment by evgeni** [comment]
Hah, the python3-libcomps from centos does not have am egg info.

**Comment by evgeni** [comment]
Yepp, only 0.1.12 got an egg: https://github.com/rpm-software-management/libcomps/commit/972ba0a75d4cdcbb5d89aa5994b3719a0c4130d5

We could artificially bump the dep to .12 or do the same trick as we did for solv.

@ehelms preferences?

**Comment by ekohl** [comment]
Do we ship a new enough version? If so, I think I'd prefer to add a version requirement that's `>= 0.1.12`.

**Comment by ehelms** [comment]
I don't think I understand the tradeoffs. Given that, follow the pattern we
did with solv is my vote. Whatever is easiest/least Jacky.

Do we need to file an issue with libcomps project?

**Comment by evgeni** [comment]
@ehelms there is no issue with libcomps, as their newer versions provide the egg, just not the one in EL8 proper (but the one we ship).
I've opted in to bumping the dep (see https://github.com/theforeman/foreman-packaging/pull/5359/commits/89caae64822ec148f8966252ff3fbc56b185b2ce) as that felt less jacky than deleting dependencies ;)

**Comment by evgeni** [comment]
And now it's modulemd on EL8, ffs.

**Comment by evgeni** [comment]
because, you know, it works in my container:

```console
[root@315aeea31b8a /]# python3
Python 3.6.8 (default, May 21 2019, 23:51:36) 
[GCC 8.2.1 20180905 (Red Hat 8.2.1-3)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import gi
>>> gi.require_version('Modulemd', '2.0')
>>> 
```

**Comment by ehelms** [comment]
GREEEN!

File: spec/setup_acceptance_node.pp [hunk 3]
```diff
@@ -6,6 +6,10 @@
       ensure => installed,
     }
 
+    package { 'epel-release':
```
Comment (ehelms) [inline spec/setup_acceptance_node.pp:9]: Is EPEL still needed for these tests to pass?

Comment (evgeni) [reply spec/setup_acceptance_node.pp:9]: Yes, we need libmodulemd2 from there.

**Pull request merged**